File: go/version_test.go
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

// The Go `const VERSION` must equal ts/package.json "version".
//
// ts/test/version.test.ts checks the TypeScript export against the SAME
// file, so the two runtimes cannot drift apart either. There is no skip
// path: an unreadable package.json FAILS here, because a version check
// that silently does not run is the exact failure mode it exists to
// prevent.

package tabnaschess

import (
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"testing"
)

func TestVersionMatchesPackageJSON(t *testing.T) {
	path := filepath.Join("..", "ts", "package.json")
	raw, err := os.ReadFile(path)
	if nil != err {
		t.Fatalf("cannot read %s, so VERSION cannot be checked: %v", path, err)
	}

	var pkg struct {
		Name    string `json:"name"`
		Version string `json:"version"`
	}
	if err := json.Unmarshal(raw, &pkg); nil != err {
		t.Fatalf("cannot parse %s: %v", path, err)
	}
	if "" == pkg.Version {
		t.Fatalf("%s has no version field, so VERSION cannot be checked", path)
	}

	if VERSION != pkg.Version {
		t.Fatalf("VERSION drift: %s exports %s but package.json is %s. "+
			"Both are rewritten at release; if you bumped one by hand, bump the other.",
			pkg.Name, VERSION, pkg.Version)
	}
}

func TestVersionIsSemver(t *testing.T) {
	if !regexp.MustCompile(`^\d+\.\d+\.\d+`).MatchString(VERSION) {
		t.Fatalf("VERSION %q must be a semver", VERSION)
	}
}

File: rs/tests/version_test.rs
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

//! `VERSION`, the crate version and `ts/package.json` "version" must all
//! agree.
//!
//! `ts/test/version.test.ts` and `go/version_test.go` check their own
//! runtimes against the SAME file, so none of the three can drift apart.
//! There is no skip path: an unreadable package.json FAILS here, because
//! a version check that silently does not run is the exact failure mode
//! it exists to prevent.

use std::fs;
use std::path::Path;

use tabnas_chess::VERSION;

#[test]
fn version_matches_cargo_toml() {
    assert_eq!(
        env!("CARGO_PKG_VERSION"),
        VERSION,
        "VERSION drift inside the crate: Cargo.toml and lib.rs disagree"
    );
}

#[test]
fn version_matches_package_json() {
    let path = Path::new("..").join("ts").join("package.json");
    let raw = fs::read_to_string(&path).unwrap_or_else(|error| {
        panic!(
            "cannot read {}, so VERSION cannot be checked: {error}",
            path.display()
        )
    });
    let package: serde_json::Value = serde_json::from_str(&raw)
        .unwrap_or_else(|error| panic!("cannot parse {}: {error}", path.display()));
    let version = package["version"].as_str().unwrap_or_else(|| {
        panic!(
            "{} has no version field, so VERSION cannot be checked",
            path.display()
        )
    });

    assert_eq!(
        version,
        VERSION,
        "VERSION drift: {} exports {VERSION} but package.json is {version}. \
         All three are rewritten at release; if you bumped one by hand, bump the others.",
        package["name"].as_str().unwrap_or("this package")
    );
}

#[test]
fn version_is_semver() {
    let mut field = VERSION.split('.');
    for part in 0..3 {
        let value = field.next().unwrap_or("");
        assert!(
            !value.is_empty() && value.bytes().all(|byte| byte.is_ascii_digit()),
            "VERSION {VERSION:?} must be a semver (field {part} is {value:?})"
        );
    }
}
